Chef provisioner SIGSEGV

Hi there,

I'm trying to run the chef provisioner in a new aws-instance using `terraform`. But unfortunately I get a `SIGSEGV` from the process and can't clearly understand the error.

### Terraform Version
`Terraform v0.9.8`

### Affected Resource(s)
- aws_instance
- provisioner "chef"

### Terraform Configuration Files
```hcl
resource "aws_instance" "graylog-mongodb-1" {
  ami             = "<REDACTED>"
  instance_type   = "t2.micro"
  key_name        = "<REDACTED>"
  security_groups = [ "<REDACTED>" ]
  subnet_id       = "<REDACTED>"
  tags            = { "Name" = "graylog-mongodb-1" }

  provisioner "chef" {
    run_list        = [ "role[infra-graylog-mongodb]" ],
    node_name       = "graylog-mongodb-1"
    server_url      = "<REDACTED>"
    user_name       = "<REDACTED>"
    os_type         = "linux"
    recreate_client = true
    user_key        = "${file("<REDACTED>")}"
    prevent_sudo    = false

    connection {
      type = "ssh"
      user = "<REDACTED>"
      private_key = "${file("<REDACTED>")}"
    }
  }

  connection {
    type = "ssh"
    user = "<REDACTED>"
    private_key = "${file("<REDACTED>")}"
  }
}
```

### Panic Output
https://gist.github.com/mateusduboli/e76654fd462bd8cdf6e3729f70a4aa7c

### Expected Behavior
Terraform should provide a friendly error message about the configuration.

### Actual Behavior
A SIGSEGV

### Steps to Reproduce
1. Run `terraform apply` with the `hcl` provided

### Important Factoids
The machine is not accessible from outside, I didn't configured the bastion host needed for it to be accessible.

I just tried to provision a new node, first time with 0.9.8 on windows, i think this is the same thing:

module.legacy_site.aws_instance.mod.2: Still creating... (50s elapsed)
module.legacy_site.aws_instance.mod.2: Provisioning with 'chef'...
Error applying plan:

1 error(s) occurred:

* module.legacy_site.aws_instance.mod[2]: 1 error(s) occurred:

* unexpected EOF

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
panic: runtime error: invalid memory address or nil pointer dereference
2017/06/09 15:05:08 [DEBUG] plugin: terraform.exe: [signal 0xc0000005 code=0x0 addr=0x18 pc=0x431f5a9]
2017/06/09 15:05:08 [DEBUG] plugin: terraform.exe:
2017/06/09 15:05:08 [DEBUG] plugin: terraform.exe: goroutine 24 [running]:
2017/06/09 15:05:08 [DEBUG] plugin: terraform.exe: github.com/hashicorp/terraform/builtin/provisioners/chef.applyFn(0x8289e40, 0xc04255f0b0, 0x0, 0x0)
2017/06/09 15:05:08 [DEBUG] plugin: terraform.exe:      /opt/gopath/src/github.com/hashicorp/terraform/builtin/provisioners/chef/resource_provisioner.go:268 +0x14d9
2017/06/09 15:05:08 [DEBUG] plugin: terraform.exe: github.com/hashicorp/terraform/helper/schema.(*Provisioner).Apply(0xc042386140, 0x825f9c0, 0xc042056428, 0xc04205a9b0, 0xc042397a10, 0x0, 0x0)
2017/06/09 15:05:08 [DEBUG] plugin: terraform.exe:      /opt/gopath/src/github.com/hashicorp/terraform/helper/schema/provisioner.go:205 +0x56a
2017/06/09 15:05:08 [DEBUG] plugin: terraform.exe: github.com/hashicorp/terraform/plugin.(*ResourceProvisionerServer).Apply(0xc0422a9ae0, 0xc0421af080, 0xc0421c8748, 0x0, 0x0)
2017/06/09 15:05:08 [DEBUG] plugin: terraform.exe:      /opt/gopath/src/github.com/hashicorp/terraform/plugin/resource_provisioner.go:142 +0x198
2017/06/09 15:05:08 [DEBUG] plugin: terraform.exe: reflect.Value.call(0xc042412840, 0xc042056410, 0x13, 0x564a125, 0x4, 0xc04273ff20, 0x3, 0x3, 0x0, 0x0, ...)
2017/06/09 15:05:08 [DEBUG] plugin: terraform.exe:      /opt/go/src/reflect/value.go:434 +0x926
2017/06/09 15:05:08 [DEBUG] plugin: terraform.exe: reflect.Value.Call(0xc042412840, 0xc042056410, 0x13, 0xc04273ff20, 0x3, 0x3, 0x0, 0x0, 0xc0423301e0)
2017/06/09 15:05:08 [DEBUG] plugin: terraform.exe:      /opt/go/src/reflect/value.go:302 +0xab
2017/06/09 15:05:08 [DEBUG] plugin: terraform.exe: net/rpc.(*service).call(0xc0423510c0, 0xc042351040, 0xc04251a598, 0xc042564d00, 0xc042237880, 0x473ccc0, 0xc0421af080, 0x16, 0x473cd00, 0xc0421c8748, ...)
2017/06/09 15:05:08 [DEBUG] plugin: terraform.exe:      /opt/go/src/net/rpc/server.go:387 +0x14b
2017/06/09 15:05:08 [DEBUG] plugin: terraform.exe: created by net/rpc.(*Server).ServeCodec
2017/06/09 15:05:08 [DEBUG] plugin: terraform.exe:      /opt/go/src/net/rpc/server.go:481 +0x40b
2017/06/09 15:05:08 [DEBUG] root.legacy_site: eval: *terraform.EvalIf
2017/06/09 15:05:08 [DEBUG] root.legacy_site: eval: *terraform.EvalUndeposeState
2017/06/09 15:05:08 [DEBUG] root.legacy_site: eval: *terraform.EvalWriteDiff
2017/06/09 15:05:08 [DEBUG] root.legacy_site: eval: *terraform.EvalApplyPost
2017/06/09 15:05:08 [ERROR] root.legacy_site: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* unexpected EOF
2017/06/09 15:05:08 [ERROR] root.legacy_site: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* unexpected EOF
2017/06/09 15:05:08 [TRACE] [walkApply] Exiting eval tree: module.legacy_site.aws_instance.mod[2]
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "provisioner.chef (close)"
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "module.legacy_site.aws_route53_record.instance[1]"
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "module.legacy_site.aws_route53_record.instance[0]"
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "module.legacy_site.aws_route53_record.instance[2]"
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "module.legacy_site.null_resource.deploy_app[0]"
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "module.legacy_site.null_resource.deploy_app[2]"
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "module.legacy_site.null_resource.deploy_app[1]"
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "module.legacy_site.aws_elb_attachment.mod[2]"
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "module.legacy_site.aws_elb_attachment.mod[1]"
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "module.legacy_site.aws_elb_attachment.mod[0]"
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "module.legacy_site.aws_elb_attachment.mod[1] (destroy)"
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "module.legacy_site.aws_elb_attachment.mod[0] (destroy)"
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "provisioner.local-exec (close)"
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "provider.aws (close)"
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "module.legacy_site.null_resource.deploy_app[0] (destroy)"
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "module.legacy_site.null_resource.deploy_app[1] (destroy)"
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "provider.null (close)"
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "provisioner.remote-exec (close)"
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "meta.count-boundary (count boundary fixup)"
2017/06/09 15:05:08 [DEBUG] dag/walk: upstream errored, not walking "root"
2017/06/09 15:05:08 [TRACE] Preserving existing state lineage "008e3fbf-6996-47a2-8cd8-ae3d83587d7e"
2017/06/09 15:05:08 [DEBUG] plugin: C:\Program Files\terraform\terraform.exe: plugin process exited
2017/06/09 15:05:08 [TRACE] Preserving existing state lineage "008e3fbf-6996-47a2-8cd8-ae3d83587d7e"
2017/06/09 15:05:08 [DEBUG] Uploading remote state to S3

--- a/builtin/provisioners/chef/resource_provisioner.go
+++ b/builtin/provisioners/chef/resource_provisioner.go
@@ -256,6 +256,7 @@ func Provisioner() terraform.ResourceProvisioner {
 // TODO: Support context cancelling (Provisioner Stop)
 func applyFn(ctx context.Context) error {
 	o := ctx.Value(schema.ProvOutputKey).(terraform.UIOutput)
+	s := ctx.Value(schema.ProvRawStateKey).(*terraform.InstanceState)
 	d := ctx.Value(schema.ProvConfigDataKey).(*schema.ResourceData)
 
 	// Decode the raw config for this provisioner
@@ -265,7 +266,7 @@ func applyFn(ctx context.Context) error {
 	}
 
 	if p.OSType == "" {
-		switch t := d.State().Ephemeral.ConnInfo["type"]; t {
+		switch t := s.Ephemeral.ConnInfo["type"]; t {
 		case "ssh", "": // The default connection type is ssh, so if the type is empty assume ssh
 			p.OSType = "linux"
 		case "winrm":
@@ -285,7 +286,7 @@ func applyFn(ctx context.Context) error {
 		p.generateClientKey = p.generateClientKeyFunc(linuxKnifeCmd, linuxConfDir, linuxNoOutput)
 		p.configureVaults = p.configureVaultsFunc(linuxGemCmd, linuxKnifeCmd, linuxConfDir)
 		p.runChefClient = p.runChefClientFunc(linuxChefCmd, linuxConfDir)
-		p.useSudo = !p.PreventSudo && d.State().Ephemeral.ConnInfo["user"] != "root"
+		p.useSudo = !p.PreventSudo && s.Ephemeral.ConnInfo["user"] != "root"
 	case "windows":
 		p.cleanupUserKeyCmd = fmt.Sprintf("cd %s && del /F /Q %s", windowsConfDir, p.UserName+".pem")
 		p.createConfigFiles = p.windowsCreateConfigFiles
@@ -300,7 +301,7 @@ func applyFn(ctx context.Context) error {
 	}
 
 	// Get a new communicator
-	comm, err := communicator.New(d.State())
+	comm, err := communicator.New(s)
 	if err != nil {
 		return err
 	}

